pkg/compact/compact.go

```diff
@@ -0,0 +1,540 @@
+			return nil
+		}
+		if err := renameFile(tmpdir, dir); err != nil {
+			level.Warn(c.logger).Log("msg", "rename temp dir", "block", id, "err", err)
+			return nil
+		}
+		// The first time we see a new block, we sort it into a group. The compactor
+		// may delete a block from a group after having compacted it.
+		// This prevents re-compacting the same data. The block remain in the top-level
+		// directory until it has actually been deleted from the bucket.
+		if err := c.add(dir); err != nil {
+			level.Warn(c.logger).Log("msg", "add new block to group", "err", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return errors.Wrap(err, "retrieve bucket block metas")
+	}
+
+	// Delete all local block dirs that no longer exist in the bucket.
+	for id, meta := range local {
+		if _, ok := remote[id]; ok {
+			continue
+		}
+		if err := os.RemoveAll(filepath.Join(c.dir, id.String())); err != nil {
+			level.Warn(c.logger).Log("msg", "delete outdated block", "block", id, "err", err)
+		}
+		if err := os.RemoveAll(filepath.Join(c.dir, "groups", c.groupKey(meta), id.String())); err != nil {
+			level.Warn(c.logger).Log("msg", "delete outdated block from group", "block", id, "err", err)
+		}
+	}
+	return nil
+}
+
+func (c *Syncer) groupKey(meta *block.Meta) string {
+	return fmt.Sprintf("%x", labels.FromMap(meta.Thanos.Labels).Hash())
+}
+
+// add adds the block in the given directory to its respective compaction group.
+func (c *Syncer) add(bdir string) error {
+	meta, err := block.ReadMetaFile(bdir)
+	if err != nil {
+		return errors.Wrap(err, "read meta file")
+	}
+	h := c.groupKey(meta)
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	g, ok := c.groups[h]
+	if !ok {
+		g, err = NewGroup(
+			log.With(c.logger, "compactionGroup", h),
+			c.bkt,
+			filepath.Join(c.dir, "groups", h),
+			meta.Thanos.Labels,
+		)
+		if err != nil {
+			return errors.Wrap(err, "create compaction group")
+		}
+		c.groups[h] = g
+	}
+	return g.Add(meta)
+}
+
+// Groups returns the compaction groups created by the Syncer.
+func (c *Syncer) Groups() (res []*Group) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	for _, g := range c.groups {
+		res = append(res, g)
+	}
+	return res
+}
+
+// GarbageCollect deletes blocks from the bucket if their data is available as part of a
+// block with a higher compaction level.
+func (c *Syncer) GarbageCollect(ctx context.Context) error {
+	// Map each block to its highest priority parent. Initial blocks have themselves
+	// in their source section, i.e. are their own parent.
+	var (
+		all     = map[ulid.ULID]*block.Meta{}
+		parents = map[ulid.ULID]ulid.ULID{}
+	)
+	err := iterMetas(c.dir, func(dir string, _ ulid.ULID) error {
+		meta, err := block.ReadMetaFile(dir)
+		if err != nil {
+			return errors.Wrap(err, "read meta")
+		}
+		all[meta.ULID] = meta
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	for id, meta := range all {
+		// For each source block we contain, check whether we are the highest priotiy parent block.
```

Comment:
highest priority parent is block with the LOWEST compaction level?

Reply:
With the highest level

Reply:
ah yea - make sense